### PR TITLE
Add noisy network and distributional Q-learning to implement Rainbow

### PR DESCRIPTION
**Pull request by joneswong**
## What do these changes do?
This pr thread is identical to 
[https://github.com/ray-project/ray/pull/2701#issue-209724098](url)
As the source branch was mis-deleted, I have to reopen a pr.

### CODE REVIEW

**Review by ericl — APPROVED** [review]
Looks good. I've only taken a cursory look over the implementation for correctness, but it does learn comparably well with noisy / distributional mode enabled.

**Comment by ericl** [comment]
screenshot from 2018-08-25 13-37-02

**Comment by ericl** [comment]
Though, compared with the Rainbow paper, our numbers are (significantly?) lower with the same hyperparams:

at 8M steps ~= 32M frames
BeamRider: ~15000 vs 1805 (ours)
Breakout: ~300 vs 67 (ours)
Qbert: ~2000 vs 1639 (ours)
SpaceInvaders: ~500 vs 737 (ours)

These are also lower than reported here for dueling DDQN: https://github.com/btaba/yarlp
I'm also running an ablation experiment to DDQN as well to see how well that does. Interestingly, the very basic rllib/dqn is beating rllib/rainbow at breakout, but this could be a fluke.

**Comment by ericl** [comment]
Here are the current results:

| env  |  Basic DQN |  Dueling DDQN | Rainbow  |  Hessel et al. DQN |  Hessel et al. Rainbow |
|---|---|---|---|---|---|
|BeamRider|2724|1074|2200|~2000|~13000|
|Breakout|281|38|72|~150|~300|
|QBert|796|3517|1059|~4000|~20000|
|SpaceInvaders|629|459|700|~500|~2000|

Btw @joneswong according to the paper shouldn't we use the KL divergence as the loss for distributional dqn instead of the cross entropy?

**Comment by joneswong** [comment]
The distributional q-learning paper <A Distributional Perspective on Reinforcement Learning> uses a cross entropy loss (see the last line of pseudo code). As for Rainbow, it describe the loss of distributional Q-learning as KL. According to the results you presented above, it is necessary to try KL. I will conduct some experiments to evaluate the KL loss.

**Comment by joneswong** [comment]
one question, denoting the target distribution of Q-value as p, the online distribution as q, minimizing KL(p || q) = p*log (p/q) = p*log(p) - p*log(q), which seems equivalent to minimizing the cross entropy loss -p*log(q). Is this my misunderstanding?

**Comment by ericl** [comment]
@joneswong I just finished an ablation test run on rainbow on {qbert, spaceinvaders, breakout, beamrider}. The sample size is small, but there seem to be some patterns:

I performed both leave-one-out ablation and single optimization runs. The raw data is here: https://github.com/ray-project/rl-experiments/tree/rainbow-ablation/atari-dqn-ablation

For each optimization, I looked at the relative benefit compared to other optimizations. I used the same hyperparameters for all experiments (and for the no-noisy case, annealed to 0.01 eps in 200k timesteps).

```
Leave-one-out ablation relative benefit:
atari-dqn-ablation/no-{beamrider, breakout, spaceinvaders, qbert}
    duel:    +2 +2 +3 +3  =  +10  (possible bug in dueling nets?)
    noisy:   +3 +3 -1 +2  =  +7  (possible bug in noisy nets?)
    double:  +1 -1 +1 -1  =  +0
    distrib: -2 -3 +2 +1  =  -2
    prio:    -1 +1 -2 -2  =  -4
    nstep:   -3 -2 -3 -3  =  -11

Just one optimization relative benefit:
atari-dqn-ablation/just-{beamrider, breakout, spaceinvaders, qbert}
    distrib: +3 +3 +3 +2  = +11
    duel:    +2 +1 +2 +1  = +6
    nstep:   +1 -2 -1 +3  = +1
    double:  -1 +2 -2 -1  = -2
    noisy:   -2 -1 +1 -3  = -5 (possible bug in noisy nets?)
    prio:    -3 -3 -3 -2  = -11 (possible bug in prioritization?)

Combined relative benefit:
    distrib: +11 - -2   =  +13
    nstep:   +1  - -11  =  +12
    double:  -2  - +0   =  -2
    duel:    +6  - +10  =  -4
    prio:    -11 - -4   =  -7 (possible bug in prioritization?)
    noisy:   -5  - +7   =  -12 (likely bug in noisy nets?)
```

Overall it seems to point to some issue in the noisy net implementation, which seems to hurt in most cases. On the other hand, we can be confident that n-step and distributional DQN are probably correct!

There also might be some issue with prioritization, though I've looked at the code a couple times and I think we're applying the IS weights correctly.

**Comment by ericl** [comment]
Btw I ran all of these concurrently on a g3.16xl machine with 4 GPUs. It seems each GPU can handle about 6 concurrent trials at a time without much latency hit.

**Comment by joneswong** [comment]
For prioritized replay, such a large capacity has never made any sense to me, say that, according to my experience, a smaller one is always better. I have no idea why do the original DQN and Rainbow paper report a 1M capacity. You can find some discussions here: https://www.endtoend.ai/slides/paper/a_deeper_look_at_experience_replay.pdf 

I checked the .yaml file. The just-noisy and no-noisy are correctly configured. However, I checked the implementation of noisy net and have not found any bug. I will check this again later. Maybe take a comparison with https://github.com/google/dopamine

**Comment by ericl** [comment]
Seems like disabling noisy nets does not improve results by that much. I also tried tuning LR and the buffer size, however the improvements were only ~2x (whereas we have a 10x gap from Rainbow reference results).

It probably makes sense to benchmark against the dopamine rainbow (perhaps starting with all enhancements disabled). @joneswong would you have time to do this? I'm tied up in the next few weeks due to prelims and other deadlines.

**Comment by ericl** [comment]
Ah so here's something suspicious: in adjust_nstep, I think we always end up truncating the last few timesteps of a trajectory, since `done=true` can only happen at the end. Also I don't think we set dones correctly during the postprocessing, so done_mask does nothing.

**Comment by joneswong** [comment]
I checked the `adjust_nstep` function. It seems that we considered n-step accumulated rewards, the next observation of the next n-th step, but ignore the next n-th `terminal` flag. Do you need me to fix this one? As for dopamine rainbow, I don't think I can test its rainbow in the next two weeks due to the ddl of osim competition. I will compare with it after the osim ddl.

**Comment by ericl** [comment]
Pong is pretty high variance, are you sure it is consistent? I would try
the other Atari tuned yamls though it takes longer to run. Also, the
episode metrics shouldn't be affected since they are calculated before
postprocessing.

Also -- is the change to nstep adjust right? If the trajectory fragment
doesn't end with done, you need to drop the last few steps right ?

**Comment by joneswong** [comment]
Thanks. I will try what you mean. I reserved the last n-1 transitions
because, imo, it will NOT bring in any harm to consider 1, 2, 3 step
Q-learning at the same time. Actually, Impala doesn't drop out the boundary
transitions in their vtrace algorithm.

**Comment by ericl** [comment]
Hm yeah, keeping the last few steps seems fine then. So one weird thing about our current Rainbow is that it doesn't exceed a reward of ~80 in Breakout, whereas DQN without the improvements gets to nearly 300. But perhaps n-step is not the cause.
